File: app/src/main/java/closer/vlllage/com/closer/handler/group/PinnedMessagesHandler.java
package closer.vlllage.com.closer.handler.group;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import closer.vlllage.com.closer.handler.data.RefreshHandler;
import closer.vlllage.com.closer.handler.helpers.ActivityHandler;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.GroupMessage;
import closer.vlllage.com.closer.store.models.GroupMessage_;
import closer.vlllage.com.closer.store.models.Pin;
import closer.vlllage.com.closer.store.models.Pin_;
import io.objectbox.android.AndroidScheduler;
import io.objectbox.reactive.DataSubscription;

public class PinnedMessagesHandler extends PoolMember {

    private RecyclerView pinnedMessagesRecyclerView;
    private GroupMessagesAdapter groupMessagesAdapter;
    private DataSubscription groupMessagesSubscription;
    private DataSubscription groupMessagesActualSubscription;

    public void attach(RecyclerView pinnedMessagesRecyclerView) {
        this.pinnedMessagesRecyclerView = pinnedMessagesRecyclerView;

        pinnedMessagesRecyclerView.setLayoutManager(new LinearLayoutManager(
                $(ActivityHandler.class).getActivity(),
                LinearLayoutManager.VERTICAL,
                true
        ));

        groupMessagesAdapter = new GroupMessagesAdapter(this);
        groupMessagesAdapter.setPinned(true);
        pinnedMessagesRecyclerView.setAdapter(groupMessagesAdapter);
    }

    public void show(Group group) {
        if (groupMessagesSubscription != null) {
            $(DisposableHandler.class).dispose(groupMessagesSubscription);
        }

        if (groupMessagesActualSubscription != null) {
            $(DisposableHandler.class).dispose(groupMessagesSubscription);
        }

        $(RefreshHandler.class).refreshPins(group.getId());

        groupMessagesSubscription = $(StoreHandler.class).getStore().box(Pin.class).query()
                .equal(Pin_.to, group.getId())
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer(pins -> {
                    if (pins.isEmpty()) {
                        setGroupMessages(new ArrayList<>());
                        return;
                    }

                    List<String> ids = new ArrayList<>();

                    for (Pin pin : pins) {
                        ids.add(pin.getFrom());
                    }

                    groupMessagesActualSubscription = $(StoreHandler.class).getStore().box(GroupMessage.class).query()
                            .in(GroupMessage_.id, ids.toArray(new String[0]))
                            .build()
                            .subscribe()
                            .single()
                            .on(AndroidScheduler.mainThread())
                            .observer(this::setGroupMessages);
                    $(DisposableHandler.class).add(groupMessagesActualSubscription);
                });

        $(DisposableHandler.class).add(groupMessagesSubscription);
    }

    private void setGroupMessages(List<GroupMessage> pinnedMessages) {
        pinnedMessagesRecyclerView.setVisibility(pinnedMessages.isEmpty() ? View.GONE : View.VISIBLE);
        groupMessagesAdapter.setGroupMessages(pinnedMessages);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/PinnedMessagesHandler.kt
package closer.vlllage.com.closer.handler.group

import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.handler.data.RefreshHandler
import closer.vlllage.com.closer.handler.helpers.ActivityHandler
import closer.vlllage.com.closer.handler.helpers.DisposableHandler
import closer.vlllage.com.closer.handler.map.MapActivityHandler
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.*
import com.queatz.on.On
import io.objectbox.android.AndroidScheduler

class PinnedMessagesHandler constructor(private val on: On) {

    private lateinit var pinnedMessagesRecyclerView: RecyclerView
    private lateinit var groupMessagesAdapter: GroupMessagesAdapter
    private var disposableGroup = on<DisposableHandler>().group()

    fun attach(pinnedMessagesRecyclerView: RecyclerView) {
        this.pinnedMessagesRecyclerView = pinnedMessagesRecyclerView

        pinnedMessagesRecyclerView.layoutManager = LinearLayoutManager(
                on<ActivityHandler>().activity,
                RecyclerView.VERTICAL,
                true
        )

        groupMessagesAdapter = GroupMessagesAdapter(On(on).apply {
            use<GroupMessageHelper> {
                pinned = true
                onSuggestionClickListener = { suggestion -> on<MapActivityHandler>().showSuggestionOnMap(suggestion) }
                onEventClickListener = { event -> on<GroupActivityTransitionHandler>().showGroupForEvent(null, event) }
                onGroupClickListener =  { group1 -> on<GroupActivityTransitionHandler>().showGroupMessages(null, group1.id) }
            }
        })

        pinnedMessagesRecyclerView.adapter = groupMessagesAdapter
    }

    fun show(group: Group) {
        disposableGroup.clear()

        on<RefreshHandler>().refreshPins(group.id!!)

        on<StoreHandler>().store.box(Pin::class).query()
                .equal(Pin_.to, group.id!!)
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer { pins ->
                    if (pins.isEmpty()) {
                        setGroupMessages(listOf())
                        return@observer
                    }

                    val ids = mutableListOf<String>()

                    for (pin in pins) {
                        ids.add(pin.from!!)
                    }

                    on<StoreHandler>().store.box(GroupMessage::class).query()
                            .`in`(GroupMessage_.id, ids.toTypedArray())
                            .build()
                            .subscribe()
                            .on(AndroidScheduler.mainThread())
                            .observer { this.setGroupMessages(it) }
                            .also { disposableGroup.add(it) }
                }
                .also { disposableGroup.add(it) }
    }

    private fun setGroupMessages(pinnedMessages: List<GroupMessage>) {
        pinnedMessagesRecyclerView.visibility = if (pinnedMessages.isEmpty()) View.GONE else View.VISIBLE
        groupMessagesAdapter.setGroupMessages(pinnedMessages)
    }
}
